### PR TITLE
alert users if they have an orphaned vmi for more than an hour

### PR DESCRIPTION
**Pull request by ashleyschuett**
Signed-off-by: Ashley Schuett <aschuett@redhat.com>

**What this PR does / why we need it**:

This is an addition alert for human operators to know if they have orphaned `vmis`. This prom query check to see if there are `vmis` running on a node that there is also a `virt-handler` running on that node, if there is not a `virt-handler` running on the node for more than an hour we alert. This allows for me discoverability into a cluster and is in addition to create an kubernetes event in this same case (https://github.com/kubevirt/kubevirt/pull/4952)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #https://bugzilla.redhat.com/show_bug.cgi?id=1868099

**Special notes for your reviewer**:

**Release note**:
```release-note
Fire Prometheus Alert when a vmi is orphaned for more than an hour 
```

### CODE REVIEW

**Review by dhiller — APPROVED** [review]
/approve

File: pkg/virt-operator/resource/generate/components/crds.go
Comment (ashleyschuett) [reply pkg/virt-operator/resource/generate/components/crds.go:773]: Than you for the review and suggestions, I have added both of them!

**Comment by qinqon** [comment]
"""
/hold
until emergency CI fixes of #5207 are merged
"""

File: pkg/virt-operator/resource/generate/components/crds.go
Comment (dankenigsberg) [inline pkg/virt-operator/resource/generate/components/crds.go:772]: I don't understand the code, but I see 60m mentioned in two different places.

For testing, I would like to be able to make this alert fire much quicker. Can you explain in the PR text what I should be doing?

**Review by dankenigsberg — CHANGES_REQUESTED** [review]

File: pkg/virt-operator/resource/generate/components/crds.go
Comment (davidvossel) [reply pkg/virt-operator/resource/generate/components/crds.go:772]: Our testing at the moment mocks this situation and verifies the rule fires given a specific scenario. It's not a real world functional test where we wait 'x' amount of time for the alert to fire, but a simulation.

For now, that simulation is all we have to work with in CI, and the 60m doesn't impact the time involved with the simulation. I think the important thing is that the simulation verifies the alert formula.

File: pkg/virt-operator/resource/generate/components/crds.go
Comment (davidvossel) [inline pkg/virt-operator/resource/generate/components/crds.go:771]: @ashleyschuett can you perhaps add a comment here to explain the formula you're using here? I kind of follow it, but I'm not 100% sure I understand the intent. I think the parts where you have `(0 * something)`  confuses me the most.

Comment (sradco) [reply pkg/virt-operator/resource/generate/components/crds.go:771]: @davidvossel  Hi, The (0*something) is to be able to do a join but not affect the value. In Prometheus it all ends up to be arithmetic calculations on the value. If you only want the value of one of them , you need to make sure the others are either 0 or 1 based on the mathematical operation you are doing. 

1. First you are counting the number of virt-launcher pods in the node. 
2. Then join it (multiply by) based on the node with the number of ready virt-handler pods (use its value) joined with the number of virt-handler pods on the node(not use its value) "or" join on the number of virt-launcher pods (not use its value)  when there are no virt-handler pods on the node so that the join will not drop the record.

Comment (ashleyschuett) [reply pkg/virt-operator/resource/generate/components/crds.go:771]: @davidvossel would you like this commented in the code as well?

Comment (davidvossel) [reply pkg/virt-operator/resource/generate/components/crds.go:771]: I'm still struggling a bit. Maybe this would help me... What prevents this expression from equaling `0` when there are no launchers and no virt-handlers on a node?

Comment (sradco) [reply pkg/virt-operator/resource/generate/components/crds.go:771]: @ashleyschuett I would consider adding the calculation as a recording rule and then use it in the alert rule. The recording rule allows you to precompute frequently needed or computationally expensive expressions and save their result as a new set of time series.

Comment (sradco) [reply pkg/virt-operator/resource/generate/components/crds.go:771]: @davidvossel If there are no launchers and no virt-handlers on a node the record should drop completely. @ashleyschuett did you test this scenario?

Comment (ashleyschuett) [reply pkg/virt-operator/resource/generate/components/crds.go:771]: @sradco yes I did. As you statued the record is completely dropped. Should I add the whole query as a recording rule, or break it up into two separate recording rules for the query to be more optimal? 

```
(count by(node)(node_namespace_pod:kube_pod_info:{pod=~'virt-launcher-.*'} ) * on (node) group_left(pod) (1*(kube_pod_container_status_ready{pod=~'virt-handler-.*'} + on (pod) group_left(node) (0 * node_namespace_pod:kube_pod_info:{pod=~'virt-handler-.*'} ))) or on (node)
```

and 
```
(0 * node_namespace_pod:kube_pod_info:{pod=~'virt-launcher-.*'} )
```

File: pkg/virt-operator/resource/generate/components/crds.go
Comment (sradco) [inline pkg/virt-operator/resource/generate/components/crds.go:844]: Hi @ashleyschuett will it be possible for you to add kubevirt_ prefix to the recording rule?
There is a PR in progress to add this prefix to the other recording rules as well, https://github.com/kubevirt/kubevirt/pull/4636.

Comment (ashleyschuett) [reply pkg/virt-operator/resource/generate/components/crds.go]: @sradco thank you! updated

**Pull request merged**
